pkg/accounting/accounting.go

```diff
@@ -367,3 +430,34 @@ func (a *Accounting) NotifyPayment(peer swarm.Address, amount uint64) error {
 
 	return nil
 }
+
+func subtractI64mU64(base int64, subtracted uint64) (result int64, err error) {
+	result = base - int64(subtracted)
+
+	if subtracted > math.MaxInt64 {
+		return 0, ErrOverflow
+	}
+
+	if result > base {
+		return 0, ErrOverflow
+	}
+
+	if result == math.MinInt64 {
+		return 0, ErrOverflow
+	}
```

Comment:
Same insight valid here as in previous case

Reply:
I think that there is a difference. In the previous cast the value is got from the Balance call and here it is calculated. Can the result be valid with math.MinInt64 value if there is no overflow? My point is that I think that this check is not an appropriate overflow validation.

Reply:
so if base is -1, and substracted is maxint64, result will be minint64, and it does not fulfill previous result > base check, as minint64 < -1

it would be a valid value, however we can not maintain working with it because of the following -1 * operations

in practice in reserve() we immediately after take -1 * the result 
in credit() we have 2 usages, one has -1 * operation following, the other would be stored as balance() which we prohibit for this value thereby
in notifypayment() we would also store this value as balance, which we prohibit now

so in settle we also check for this when reading balance() which seems redundant at first glance as we dont allow saving this value previously.
only because we have to convert negative times this value to uint64() which would produce a non correct value to settle for because of the possible overflow

Reply:
> so if base is -1, and substracted is maxint64, result will be minint64, and it does not fulfill previous result > base check, as minint64 < -1

Should we then have a check `if base == -1 && substracted == math.MaxInt64`, as this is the case that needs to be handled, as there are other arithmetic operations that can produce the `result == math.MinInt64`?

> 
> it would be a valid value, however we can not maintain working with it because of the following -1 * operations

It is a valid value here, but the following functions should handle that value. Overflow error here is false positive in a set of operations.

> 
> in practice in reserve() we immediately after take -1 * the result
> in credit() we have 2 usages, one has -1 * operation following, the other would be stored as balance() which we prohibit for this value thereby
> in notifypayment() we would also store this value as balance, which we prohibit now

These other functions rely on the behaviour of this function in non-safe way. If someone would remove this check here, the consequence is that the problem would be silently reintroduced. This makes the code hard to maintain in time. There should be the check for valid values as close to the root cause as as clear as possible.

> 
> so in settle we also check for this when reading balance() which seems redundant at first glance as we dont allow saving this value previously.
> only because we have to convert negative times this value to uint64() which would produce a non correct value to settle for because of the possible overflow

For this case, there should be a comment explaining that this check is in the purpose of historical data handling.

Reply:
yes, base can be any negative int64, so there is a positive int64 that can be subtracted so that it creates the minint64 result

so this function has 2 purposes,
making a safe int64 - uint64 operation;
filtering results that can not be represented in int64 if multiplied by -1

it can be argued that this second purpose is not really necessary to use this when its only about whether we'd store minInt64 value

so we could have 2 versions of this function, 1 with this check, 1 without this check, but we definietly need this check when we are going to use -1 times the result. perhaps the version with the -1 safety check should involve the return -1 * result so it's usage is never mixed by accident with the one which is not intended to be used as -1 * afterwards

Reply:
on the other hand, if we have this value stored, than settle will just always fail with overflow error in it's current form, so I am still in favor of keeping this check as is